pkg/controller/auditing/auditing_controller.go

```diff
@@ -0,0 +1,144 @@
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	kubespherescheme "kubesphere.io/kubesphere/pkg/client/clientset/versioned/scheme"
+	auditinginformer "kubesphere.io/kubesphere/pkg/client/informers/externalversions/auditing/v1alpha1"
+	auditinglister "kubesphere.io/kubesphere/pkg/client/listers/auditing/v1alpha1"
+	"time"
+)
+
+type Controller struct {
+	auditingLister auditinglister.WebhookLister
+	auditingSynced cache.InformerSynced
+	workQueue      workqueue.RateLimitingInterface
+}
+
+func NewController(informer auditinginformer.WebhookInformer) *Controller {
+	// Create auditing webhook informer
+
+	utilruntime.Must(kubespherescheme.AddToScheme(scheme.Scheme))
+
+	return &Controller{
+		auditingLister: informer.Lister(),
+		auditingSynced: informer.Informer().HasSynced,
+		workQueue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "auditing"),
+	}
+}
+
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workQueue.ShutDown()
+
+	//init client
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting User controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.auditingSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Info("Starting workers")
+	// Launch two workers to process Foo resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	klog.Info("Started workers")
+	<-stopCh
+	klog.Info("Shutting down workers")
+	return nil
+}
+
+func (c *Controller) enqueueUser(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	c.workQueue.Add(key)
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workQueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the workqueue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer c.workQueue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if key, ok = obj.(string); !ok {
```

Comment:
What does this auditing controller do? This is no reconciling work to do.

Reply:
In 3.0, user can only edit the CRD Webhook and Rule with yaml, the usage of controller is get CRD Webhook. But in 3.1, it will be allow user to custom Webhook and Rule in UI.

Reply:
But, there is nothing to be done here in controller. Which part is necessary for CRD webhook?

Reply:
Now it only used to get the auditlevel and k8s audit enabled status from CRD Webhook, auditlevel  controlles the audit event record level, and k8s audit enabled status determins ks-apiserver is auditing the request to kube-apiserver wether or not. Ks-apiserver need to get these two parameter in every request, so it need a controller.

Reply:
The controller has no use in 3.0, i will delete it and add in 3.1.

Reply:
> Now it only used to get the auditlevel and k8s audit enabled status from CRD Webhook, auditlevel controlles the audit event record level, and k8s audit enabled status determins ks-apiserver is auditing the request to kube-apiserver wether or not. Ks-apiserver need to get these two parameter in every request, so it need a controller.

auditlevel and K8sAuditEnabled status can be get by a lister to the webhook CRD, a controller is not necessary maybe? Why we need a controller here?

K8sAuditEnabled should only reflect K8s audit is on or off. It should not be related to KubeSphere audit status(on or off)

Reply:
Now the controller has be removed, there only has a informer to get the CRD Webhook.
K8sAuditEnabled is only reflect K8s audit is on or off, not to KubeSphere audit.
Ks audit level can be modified by modified CRD Webhook,  k8s audit level can be modified by modified the audit policy file when use static audit or modified the AuditSink when use dynamic audit.

Reply:
> Now the controller has be removed, there only has a informer to get the CRD Webhook.
> K8sAuditEnabled is only reflect K8s audit is on or off, not to KubeSphere audit.
> Ks audit level can be modified by modified CRD Webhook, k8s audit level can be modified by modified the audit policy file when use static audit or modified the AuditSink when use dynamic audit.

That's sounds good